base_layer/core/src/mining/miner.rs

```diff
@@ -114,32 +119,32 @@ impl<B: BlockchainBackend> Miner<B> {
     /// 4. We iterate on the header nonce until
     ///     * the target difficulty is reached
     ///     * or the loop is interrupted because a new block was found in the interim, or the miner is stopped
-    async fn mining(&mut self) -> Result<(), MinerError> {
+    async fn mining(mut self) -> Result<Miner<B>, MinerError> {
         // Lets make sure its set to mine
-        debug!(target: LOG_TARGET, "Start mining thread");
-        while !self.kill_flag.load(Ordering::Relaxed) {
-            while !self.received_new_block_flag.load(Ordering::Relaxed) {
-                tokio::time::delay_for(Duration::from_millis(100)).await; // wait for new block event
-                if self.kill_flag.load(Ordering::Relaxed) {
-                    debug!(target: LOG_TARGET, "Mining stopped with kill flag.");
-                    return Ok(());
-                }
-            }
-            let flag = self.received_new_block_flag.clone();
-            flag.store(false, Ordering::Relaxed);
-            debug!(target: LOG_TARGET, "Miner asking for new candidate block to mine.");
-            let mut block_template = self.get_block_template().await?;
-            let output = self.add_coinbase(&mut block_template)?;
-            let mut block = self.get_block(block_template).await?;
-            debug!(target: LOG_TARGET, "Miner got new block to mine.");
-            let difficulty = self.get_req_difficulty().await?;
-            let new_block_event_flag = self.received_new_block_flag.clone();
-            let kill = self.kill_flag.clone();
+        debug!(target: LOG_TARGET, "Miner asking for new candidate block to mine.");
+        let mut block_template = self.get_block_template().await?;
+        let output = self.add_coinbase(&mut block_template)?;
+        let mut block = self.get_block(block_template).await?;
+        debug!(target: LOG_TARGET, "Miner got new block to mine.");
+        let difficulty = self.get_req_difficulty().await?;
+        let (tx, mut rx): (Sender<Option<BlockHeader>>, Receiver<Option<BlockHeader>>) = mpsc::channel(self.threads);
+        for _ in 0..self.threads {
+            let stop_mining_flag = self.stop_mining_flag.clone();
             let header = block.header.clone();
-            let mining_handle =
-                spawn_blocking(move || CpuBlakePow::mine(difficulty, header, new_block_event_flag, kill));
-            let result = mining_handle.await.unwrap_or(None);
-            if let Some(r) = result {
+            let mut tx_channel = tx.clone();
+            trace!("spawning mining thread");
+            spawn_blocking(move || {
+                let result = CpuBlakePow::mine(difficulty, header, stop_mining_flag);
+                // send back what the miner found, None will be sent if the miner did not find a nonce
+                tx_channel.try_send(result);
+            });
+        }
+        drop(tx); // lets ensure that the tx all get dropped
+        'process: while let Some(value) = rx.next().await {
```

Comment:
You don't need to label this loop - IMO loop labels are code smell and never actually needed